File: host-libs/js/loaders/python.ts
let _lib: Deno.DynamicLibrary<typeof PYTHON_SYMBOLS> | null = null;

const PYTHON_SYMBOLS = {
    polyplug_python_loader_create: {
        parameters: ["pointer"] as const,
        result: "pointer" as const,
    },
} satisfies Deno.ForeignLibraryInterface;

function getLib() {
    if (!_lib) {
        const libPath = Deno.env.get("POLYPLUG_PYTHON_LIB")
            ?? "libpolyplug_python.so";
        _lib = Deno.dlopen(libPath, PYTHON_SYMBOLS);
    }
    return _lib;
}

export function registerPythonLoader(
    rt: Deno.PointerValue,
    registerFn: (rt: Deno.PointerValue, loader: Deno.PointerValue) => number,
    minVersion: string = "3.11"
): void {
    const lib = getLib();
    const encoded = new TextEncoder().encode(minVersion);
    const cfgBuf = new Uint8Array(16); // ptr (8) + len (8)
    const ptr = Deno.UnsafePointer.of(encoded);
    const view = new DataView(cfgBuf.buffer);
    view.setBigUint64(0, BigInt(Deno.UnsafePointer.value(ptr)), true);
    view.setBigUint64(8, BigInt(encoded.length), true);
    const cfgPtr = Deno.UnsafePointer.of(cfgBuf);
    const loaderPtr = lib.symbols.polyplug_python_loader_create(cfgPtr);
    if (loaderPtr === null) {
        throw new Error("polyplug: python loader create failed");
    }
    const err = registerFn(rt, loaderPtr);
    if (err !== 0) {
        throw new Error(`polyplug: python loader register failed: ${err}`);
    }
}

File: host-libs/lua/loaders/polyplug-loaders-python/polyplug/loaders/python.lua
--- Python loader registration for polyplug.
local ffi = require("ffi")

local ok = pcall(ffi.cdef, [[
    typedef struct { const uint8_t* ptr; size_t len; } PolyplugPythonConfig;
    void* polyplug_python_loader_create(const PolyplugPythonConfig* cfg);
    void  polyplug_python_loader_free(void* ptr);
    uint32_t polyplug_runtime_register_loader(void* rt, void* loader);
]])

local _lib = nil
local function get_lib()
    if not _lib then
        _lib = ffi.load("polyplug_python")
    end
    return _lib
end

local M = {}

function M.register(rt, min_version)
    min_version = min_version or "3.11"
    local lib = get_lib()
    local cfg = ffi.new("PolyplugPythonConfig", {
        ffi.cast("const uint8_t*", min_version),
        #min_version
    })
    local loader = lib.polyplug_python_loader_create(cfg)
    if loader == nil then
        error("polyplug: python loader create failed")
    end
    local polyplug_lib = ffi.load("polyplug")
    local err = polyplug_lib.polyplug_runtime_register_loader(rt, loader)
    if err ~= 0 then
        error("polyplug: python loader register failed: " .. tostring(err))
    end
end

return M